src/utils/py/README.md

```diff
@@ -1 +1,244 @@
+```python
+class DaosContext(object)
+class DaosLog
+class DaosApiError(Exception)
+```
+
+`DaosContext` is a wrapper around the DAOS libraries. It is initialized with the path where DAOS libraries can be found. It describes the environmental context in which to run.
+
+`DaosLog` exposes functionality to write messages to the DAOS client log.
+
+`DaosApiError` is a custom exception class raised by the API internally in the event of a failed DAOS action.
+
+Every function exposed in the DAOS C API supports both synchronous and asynchronous execution, and the Python API exposes this same functionality. Each API takes an input event. `DaosEvent` is the Python representation of this event. If the input event is `NULL`, the call is synchronous. If an event is supplied, the function will return immediately after submitting API requests to the underlying stack and the user can poll and query the event for completion.
+
+### Ctypes
+
+Ctypes is a built-in Python module for interfacing Python with existing libraries written in C/C++. The Python API is built as an object-oriented wrapper around the DAOS libraries utilizing ctypes.
+
+Ctypes documentation can be found here <https://docs.python.org/3/library/ctypes.html>
+
+The following demonstrates a simplified example of calling the C function `daos_pool_tgt_exclude_out` from Python, with each input parameter to the C function being cast via ctypes. This also demonstrates struct representation via ctypes:
+
+```C
+// daos_exclude.c
+
+#include <stdio.h>
+
+int
+daos_pool_tgt_exclude_out(const uuid_t uuid, const char *grp,
+                          const d_rank_list_t *svc, struct d_tgt_list *tgts,
+                          daos_event_t *ev);
+```
+
+All input parameters must be represented via ctypes. If a struct is required as an input parameter, a corresponding Python class can be created. For struct `d_tgt_list`:
+
+```c
+struct d_tgt_list {
+	d_rank_t	*tl_ranks;
+	int32_t		*tl_tgts;
+	uint32_t	tl_nr;
+};
+```
+```python
+class DTgtList(ctypes.Structure):
+    _fields_ = [("tl_ranks", ctypes.POINTER(ctypes.c_uint32)),
+                ("tl_tgts", ctypes.POINTER(ctypes.c_int32)),
+                ("tl_nr", ctypes.c_uint32)]
+```
+
+The shared object containing `daos_pool_tgt_exclude_out` can then be imported and the function called directly:
+
+```python
+# api.py
+
+import ctypes
+import uuid
+import conversion # utility library to convert C <---> Python UUIDs
+
+# init python variables
+p_uuid = str(uuid.uuid4())
+p_tgts = 2
+p_ranks = DaosPool.__pylist_to_array([2])
+
+# cast python variables via ctypes as necessary
+c_uuid = str_to_c_uuid(p_uuid)
+c_grp = ctypes.create_string_buffer(b"daos_group_name")
+c_svc = ctypes.POINTER(2) # ensure pointers are cast/passed as such
+c_tgt_list = ctypes.POINTER(DTgtList(p_ranks, p_tgts, 2))) # again, DTgtList must be passed as pointer
+
+# load the shared object
+my_lib = ctypes.CDLL('/full/path/to/daos_exclude.so')
```

Comment:
This is a little confusing since the context handles it and no one would ever need to do this.

Reply:
Giving this a bit more thought, I do like this here rather than showing creating DaosContext, etc. The intent for this section is to show ctypes usage and someone calling a C function would need to load the .so to call it. The Python API handles this internally in DaosContext, but this is demonstrated below where a user of the API would be using it, in the Usage section.